implementations/rust/ockam/ockam_core/src/error/inner/mod.rs

```diff
@@ -0,0 +1,183 @@
+    #[track_caller]
+    #[cfg(not(feature = "std"))]
+    pub(super) fn new_without_cause(origin: Origin, kind: Kind) -> Self {
+        Self {
+            code: ErrorCode::new(origin, kind),
+        }
+        // Self::new_inner(ErrorCode::new(origin, kind), None, None)
+    }
+
+    #[cold]
+    #[track_caller]
+    #[cfg(feature = "std")]
+    pub(super) fn new_inner(
+        code: ErrorCode,
+        cause: Option<Box<dyn ErrorTrait + Send + Sync>>,
+        type_name: Option<&'static str>,
+    ) -> Self {
+        let location = core::panic::Location::caller();
+        #[allow(unused_mut)]
+        let mut local: Vec<LocalPayloadEntry> = vec![];
+
+        #[cfg(all(feature = "std", feature = "error-traces"))]
+        if trace_config::BACKTRACE_ENABLED.get() {
+            local.push(LocalPayloadEntry::Backtrace(backtrace::Backtrace::new()));
+        }
+
+        #[cfg(all(feature = "std", feature = "error-traces"))]
+        if trace_config::SPANTRACE_ENABLED.get() {
+            local.push(LocalPayloadEntry::Spantrace(
+                tracing_error::SpanTrace::capture(),
+            ));
+        }
+
+        let mut payload = vec![];
+        if let Some(cause) = cause.as_ref() {
+            let debug = crate::compat::format!("{:?}", cause);
+            let display = crate::compat::format!("{}", cause);
+            payload.push(PayloadEntry::Cause {
+                display,
+                debug,
+                type_name: type_name.unwrap_or("<unknown>").to_owned(),
+            });
+        }
+
+        Self {
+            code,
+            cause,
+            source_loc: location.into(),
+            local: Some(local),
+            payload,
+        }
+    }
+
+    #[cold]
+    pub fn add_context(&mut self, _key: &str, _val: &dyn core::fmt::Display) {
+        #[cfg(feature = "std")]
+        self.payload.push(PayloadEntry::Info(
+            _key.into(),
+            crate::compat::format!("{}", _val),
+        ));
+    }
+
+    #[cfg(feature = "std")]
+    pub fn cause(&self) -> Option<&(dyn ErrorTrait + Send + Sync + 'static)> {
+        self.cause.as_deref()
+    }
+
+    #[cfg(not(feature = "std"))]
+    pub fn cause(&self) -> Option<&(dyn ErrorTrait + Send + Sync + 'static)> {
+        None
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+enum PayloadEntry {
+    // `Display` for the error cause (note that this may instead be used for the
+    // `message`).
+    Cause {
+        display: String,
+        debug: String,
+        type_name: String,
+    },
+    // Miscellaneous info — type name of the worker that caused the error for
+    // example, or a human description of the cause.
+    Info(String, String),
+}
+
+// Information that is only relevant on this machine. May be logged, so should
+// not contain sensitive data, but is discarded when serializing.
+#[derive(Debug)]
+enum LocalPayloadEntry {
+    #[cfg(all(feature = "std", feature = "backtrace"))]
+    Backtrace(backtrace::Backtrace),
```

Comment:
How is that different from backtrace inside `cause` field?

Reply:
You mean in the dyn Error? https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.backtrace just the interface by which you can return a backtrace. It is not automatically provided or anything like that, you still have to capture it.